File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
arr = ['A', 'E', 'I', 'O', 'U']
lst = []
def choose(n, dic):
    if len(lst) == n:
        dic.append(''.join(lst))
        return

    for w in arr:
        lst.append(w)
        choose(n, dic)
        lst.pop()

def solution(word):
    dic = []

    for i in range(1, 6):
        choose(i, dic)
    
    return sorted(dic).index(word)+1

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    static List<String> result = new ArrayList<>();
    
    public int solution(String word) {
        // A, E, I, O, U 로 1부터 5까지의 r을 조합
        // 정렬 후 word의 index + 1
        String[] arr = new String[]{"A", "E", "I", "O", "U"};
        
        for (int r = 1; r <= 5; r++) {
            per(arr, 5, r, "");
        }
        
        result = result.stream()
            .sorted()
            .toList();
        
        return result.indexOf(word) + 1;
    }
    
    public void per(String[] arr, int n, int r, String s) {
        if (s.length() == r) {
            result.add(s);
            return;
        }
        
        for (int i = 0; i < arr.length; i++) {
            per(arr, n, r, s+arr[i]);
        }
    }
}